src/mgr/Action.cc

```diff
@@ -111,12 +111,35 @@ Mgr::Action::fillEntry(StoreEntry* entry, bool writeHttpHeader)
 
         entry->replaceHttpReply(rep);
     }
-
+    if (!aggregatable())
+        openKidSection(entry, is_yaml());
     dump(entry);
+    if (!aggregatable() && atomic())
+        closeKidSection(entry, is_yaml());
 
     entry->flush();
 
     if (atomic())
         entry->complete();
 }
 
+void
+Mgr::openKidSection(StoreEntry *entry, const bool is_yaml)
+{
+    if (!UsingSmp())
+        return;
```

Comment:
Er.. I'm confused. The spec says that a YAML map with a "kid" section; it says nothing about who should emit that "kid" entry. The current code places that burden on the infrastructure for non-aggregatable jobs. Why should we change that?

Reply:
> The spec ... says nothing about who should emit that "kid" entry. The current code places that burden on the infrastructure for non-aggregatable jobs. Why should we change that?

Hm... I am having trouble connecting the question with the change request. This change request does not change _who_ emits the kid entry. It changes _when_ the kid entry is emitted. After this change request is addressed, for YAML reports, the entry will be emitted in more cases (to comply with specs v2) because it will no longer be silenced for non-SMP cases.

In other words, producers of non-aggregated content will call openKidSection() and emit the "kid" entry in both SMP and non-SMP modes. AFAICT, this matches specs v2 partially quoted below:

> 1: A kid-specific document produced by a kid process is a YAML map containing a `kid: ID` entry ...
> 2: A document produced by Coordinator by aggregating SMP kid information is a YAML map containing a `kids: [ ... ]` entry

Should we reorder the above items and make it more explicit that they are mutually exclusive? Or is the problem elsewhere?

Reply:
yes. But before I added the setUris() call, it would fail an Assure check that the well-known-prefix be present. Now it is working, although it's noisy because Assure uses ReportAndThrow() which unconditionally prints to standard error in unit tests even when the resulting exception is caught

Reply:
If Assure() throws during a test, then that test is not working correctly (unless the test is specifically designed to trigger Squid assertions). Assure() is an assertion -- an indication of a Squid bug.

Reply:
They do. See testCacheManager.cc:218

Reply:
> Francesco: Right now the main effect is to break testCacheMgr: testRegister() creates ...

> Alex: PR diff for testRegister() shows explicit mem_obj creation ...

> Francesco: yes. But before I added [that code] it would fail an Assure check

> Alex: If Assure() throws during a test, then that test is not working correctly (unless the test is specifically designed to trigger Squid assertions). Assure() is an assertion -- an indication of a Squid bug.

> Francesco: They do.

"They"? There is only one plural in my statement -- "assertions". "Assertions do" does not really make sense in this context.

> See testCacheManager.cc:218

I am aware of testInvalidUrl(). How is that testParseUrl()-specific code related to testRegister() we were discussing?